functions.php

```diff
@@ -0,0 +1,270 @@
+
+function readProducts(string $path): ?array
+{
+    $productsRead = @file_get_contents($path);
+
+    if ($productsRead === false) {
+        return null;
+    }
+
+    return fromJson($productsRead);
+}
+
+function extractProduct(array $products, string $category, ?int $id = null): ?array
+{
+    if (!isset($products[$category])) {
+        return null;
+    }
+
+    if ($id !== null) {
+        foreach ($products[$category] as $product) {
+            if ($product['id'] == $id) {
+                return $product;
+            }
+        }
+
+        return null;
+    }
+
+    return $products[$category];
+}
+
+function addToList(array &$products, array $requestData, ?array $oldProduct = null, bool $newId = false): void
+{
+    $fields = [ID, NAME, PRICE, QUANTITY];
+
+    foreach ($fields as $property) {
+        $updatedProduct[$property] = (
+            isset($requestData[$property]) ? $requestData[$property] :
+            (isset($oldProduct[$property]) ? $oldProduct[$property] : null)
+        );
+    }
+
+    if (!isset($products[$requestData[CATEGORY]])) {
+        $products[$requestData[CATEGORY]] = [];
+    }
+
+    if ($newId) {
+        $updatedProduct[ID] = generateId($products[$requestData[CATEGORY]], ID);
+        $products[$requestData[CATEGORY]][] = $updatedProduct;
+    } else {
+        foreach($products[$requestData[CATEGORY]] as &$product) {
+            if ($product[ID] === $updatedProduct[ID]) {
+                $product = $updatedProduct;
+            }
+        }
+    }
+}
+
+function generateId(array $categoryOfProducts, string $idKeyName): int
+{
+    $id = 0;
+
+    if (count($categoryOfProducts) > 0) {
+        foreach ($categoryOfProducts as $product) {
+            if ($product[$idKeyName] >= $id) {
+                $id = $product[$idKeyName] + 1;
+            }
+        }
+    }
+
+    return $id;
+}
+
+function validateRequestData(array $mandatoryFields, ?array $requestData = null): void
+{
+    if ($requestData === null) {
+        sendResponse(toJson(['error' => 'Sent data is not correct']), 414);
+    }
+
+    $missingFields = array_keys(array_diff_key($mandatoryFields, $requestData));
+
+    if ($missingFields) {
+        sendResponse(toJson(['error' => 'Missing fields: ' . implode(', ', $missingFields)]), 400);
+    }
+
+    $typeErrors = validateType($mandatoryFields, $requestData);
+
+    if($typeErrors) {
+        sendResponse(toJson(['error' => implode(', ', $typeErrors)]), 500);
+    }
+}
+
+function validateType(array $rules, array $requestData, bool $optional = false): ?array
+{
+    $errors = [];
+
+    foreach ($rules as $key => $value) {
+        switch ($value) {
+            case 'int':
+                if ($optional && isset($requestData[$key]) && !is_int($requestData[$key])) {
+                    $errors[] = "Key {$key} must be {$value}";
+                } elseif (!$optional && !is_int($requestData[$key])) {
+                    $errors[] = "Key {$key} must be {$value}";
+                }
```

Comment:
```suggestion
                if ($optional && !isset($requestData[$key])) {
                    continue;
                }

                if (!$optional && !isset($requestData[$key])) {
                    $errors[] = "Key {$key} is required";
                }

                if (!is_int($requestData[$key])) {
                    $errors[] = "Key {$key} must be {$value}";
                }
```

Reply:
@AndreiOrmanji - I think the use of `continue` in a switch is equivalent to `break` .

As the logic of the `validateType()` is to check the type of the fields in array and in situations of when `$optional` flag is used then to check the type only if it exists in the array, I believe this change would be a better approach: 

```
                if ($optional && !isset($requestData[$key])) {
                    break;
                }

                if (!is_int($requestData[$key])) {
                    $errors[] = "Key {$key} must be {$value}";
                }
                break;
```

Reply:
@LabiciIvan, ok, use your variant here.